### PR TITLE
Refactor processing return data from network.load_config

### PR DESCRIPTION
**Pull request by bewing**
### What does this PR do?
Move default_ret and loaded_ret into salt.utils.napalm, and update the
dictionary returned based on discussions in #40208. Diffs are copied
into comment if this is a dry-run, and changes are only set if an
actual change is made on the device.

### Previous Behavior
```yaml
[root@saltmaster:~] # salt --out=yaml veos1 state.apply test=True
veos1:
  netacl_|-test_|-test_|-managed:
    __id__: test
    __run_num__: 1
    __sls__: router.mcast.acl
    already_configured: false
    changes:
      diff: "@@ -26,6 +26,18 @@\n interface Management1\n    ip address 10.0.2.15/24\n
        !\n+ip access-list standard a-mcast\n+   10 remark $Id: test $\n+   20
        remark term1\n+   30 permit host 233.252.21.1\n+   40 permit host 233.252.21.2\n+!\n+ip
        access-list standard b-mcast\n+   10 remark $Id: test $\n+   20 remark
        term1\n+   30 permit host 233.252.21.129\n+   40 permit host 233.252.21.130\n+!\n
        ip route 0.0.0.0/0 10.0.2.2\n !\n no ip routing"
    comment: 'Testing mode: Configuration discarded.'
    duration: 2490.159
    name: test
    result: null
    start_time: '19:04:19.338984'
  netconfig_|-mcast-base_|-mcast-base_|-managed:
    __id__: mcast-base
    __run_num__: 0
    __sls__: router.mcast
    already_configured: true
    changes:
      diff: ''
    comment: 'Testing mode: Configuration discarded.'
    duration: 1699.846
    loaded_config: ''
    name: mcast-base
    result: null
    start_time: '19:04:17.637968'
  netconfig_|-pim-rps_|-pim-rps_|-managed:
    __id__: pim-rps
    __run_num__: 2
    __sls__: router.mcast.pim
    already_configured: false
    changes:
      diff: "@@ -32,6 +32,10 @@\n !\n ip multicast-routing\n !\n+router pim sparse-mode\n+
        \  ip pim rp-address 192.0.2.1 access-list a-mcast\n+   ip pim rp-address
        192.0.2.2 access-list b-mcast\n+!\n management api http-commands\n    no
        shutdown\n !"
    comment: 'Testing mode: Configuration discarded.'
    duration: 1958.6
    loaded_config: ''
    name: pim-rps
    result: null
    start_time: '19:04:21.829395'
[root@saltmaster:~] # salt --out=yaml veos1 state.apply
veos1:
  netacl_|-test_|-test_|-managed:
    __id__: test
    __run_num__: 1
    __sls__: router.mcast.acl
    already_configured: false
    changes:
      diff: "@@ -26,6 +26,18 @@\n interface Management1\n    ip address 10.0.2.15/24\n
        !\n+ip access-list standard a-mcast\n+   10 remark $Id: test $\n+   20
        remark term1\n+   30 permit host 233.252.21.1\n+   40 permit host 233.252.21.2\n+!\n+ip
        access-list standard b-mcast\n+   10 remark $Id: test $\n+   20 remark
        term1\n+   30 permit host 233.252.21.129\n+   40 permit host 233.252.21.130\n+!\n
        ip route 0.0.0.0/0 10.0.2.2\n !\n no ip routing"
    comment: Configuration changed!
    duration: 2524.962
    name: test
    result: true
    start_time: '19:04:43.764291'
  netconfig_|-mcast-base_|-mcast-base_|-managed:
    __id__: mcast-base
    __run_num__: 0
    __sls__: router.mcast
    already_configured: true
    changes:
      diff: ''
    comment: Already configured.
    duration: 2094.271
    loaded_config: ''
    name: mcast-base
    result: true
    start_time: '19:04:41.668861'
  netconfig_|-pim-rps_|-pim-rps_|-managed:
    __id__: pim-rps
    __run_num__: 2
    __sls__: router.mcast.pim
    already_configured: false
    changes:
      diff: "@@ -44,6 +44,10 @@\n !\n ip multicast-routing\n !\n+router pim sparse-mode\n+
        \  ip pim rp-address 192.0.2.1 access-list a-mcast\n+   ip pim rp-address
        192.0.2.2 access-list b-mcast\n+!\n management api http-commands\n    no
        shutdown\n !"
    comment: Configuration changed!
    duration: 2615.863
    loaded_config: ''
    name: pim-rps
    result: true
    start_time: '19:04:46.289470'
```
### New Behavior
```yaml
[root@saltmaster:~] # salt --out=yaml veos1 state.apply test=True
veos1:
  netacl_|-test_|-test_|-managed:
    __id__: test
    __run_num__: 1
    __sls__: router.mcast.acl
    changes: {}
    comment: "To be changed: @@ -26,6 +26,18 @@\n interface Management1\n    ip address
      10.0.2.15/24\n !\n+ip access-list standard a-mcast\n+   10 remark $Id: test
      $\n+   20 remark term1\n+   30 permit host 233.252.21.1\n+   40 permit host
      233.252.21.2\n+!\n+ip access-list standard b-mcast\n+   10 remark $Id: test
      $\n+   20 remark term1\n+   30 permit host 233.252.21.129\n+   40 permit host
      233.252.21.130\n+!\n ip route 0.0.0.0/0 10.0.2.2\n !\n no ip routing\nConfiguration
      discarded."
    duration: 2122.099
    name: test
    pchanges:
      diff: "@@ -26,6 +26,18 @@\n interface Management1\n    ip address 10.0.2.15/24\n
        !\n+ip access-list standard a-mcast\n+   10 remark $Id: test $\n+   20
        remark term1\n+   30 permit host 233.252.21.1\n+   40 permit host 233.252.21.2\n+!\n+ip
        access-list standard b-mcast\n+   10 remark $Id: test $\n+   20 remark
        term1\n+   30 permit host 233.252.21.129\n+   40 permit host 233.252.21.130\n+!\n
        ip route 0.0.0.0/0 10.0.2.2\n !\n no ip routing"
    result: null
    start_time: '18:48:58.539133'
  netconfig_|-mcast-base_|-mcast-base_|-managed:
    __id__: mcast-base
    __run_num__: 0
    __sls__: router.mcast
    changes: {}
    comment: Configuration discarded.
    duration: 1694.58
    name: mcast-base
    pchanges: {}
    result: true
    start_time: '18:48:56.843249'
  netconfig_|-pim-rps_|-pim-rps_|-managed:
    __id__: pim-rps
    __run_num__: 2
    __sls__: router.mcast.pim
    changes: {}
    comment: "To be changed: @@ -32,6 +32,10 @@\n !\n ip multicast-routing\n !\n+router
      pim sparse-mode\n+   ip pim rp-address 192.0.2.1 access-list a-mcast\n+
      \  ip pim rp-address 192.0.2.2 access-list b-mcast\n+!\n management api
      http-commands\n    no shutdown\n !\nConfiguration discarded."
    duration: 2306.993
    name: pim-rps
    pchanges:
      diff: "@@ -32,6 +32,10 @@\n !\n ip multicast-routing\n !\n+router pim sparse-mode\n+
        \  ip pim rp-address 192.0.2.1 access-list a-mcast\n+   ip pim rp-address
        192.0.2.2 access-list b-mcast\n+!\n management api http-commands\n    no
        shutdown\n !"
    result: null
    start_time: '18:49:00.661585'
[root@saltmaster:~] # salt --out=yaml veos1 state.apply
veos1:
  netacl_|-test_|-test_|-managed:
    __id__: test
    __run_num__: 1
    __sls__: router.mcast.acl
    changes:
      diff: "@@ -26,6 +26,18 @@\n interface Management1\n    ip address 10.0.2.15/24\n
        !\n+ip access-list standard a-mcast\n+   10 remark $Id: test $\n+   20
        remark term1\n+   30 permit host 233.252.21.1\n+   40 permit host 233.252.21.2\n+!\n+ip
        access-list standard b-mcast\n+   10 remark $Id: test $\n+   20 remark
        term1\n+   30 permit host 233.252.21.129\n+   40 permit host 233.252.21.130\n+!\n
        ip route 0.0.0.0/0 10.0.2.2\n !\n no ip routing"
    comment: ''
    duration: 2677.044
    name: test
    pchanges:
      diff: "@@ -26,6 +26,18 @@\n interface Management1\n    ip address 10.0.2.15/24\n
        !\n+ip access-list standard a-mcast\n+   10 remark $Id: test $\n+   20
        remark term1\n+   30 permit host 233.252.21.1\n+   40 permit host 233.252.21.2\n+!\n+ip
        access-list standard b-mcast\n+   10 remark $Id: test $\n+   20 remark
        term1\n+   30 permit host 233.252.21.129\n+   40 permit host 233.252.21.130\n+!\n
        ip route 0.0.0.0/0 10.0.2.2\n !\n no ip routing"
    result: true
    start_time: '18:49:26.936205'
  netconfig_|-mcast-base_|-mcast-base_|-managed:
    __id__: mcast-base
    __run_num__: 0
    __sls__: router.mcast
    changes: {}
    comment: Already configured.
    duration: 1720.124
    name: mcast-base
    pchanges: {}
    result: true
    start_time: '18:49:25.214744'
  netconfig_|-pim-rps_|-pim-rps_|-managed:
    __id__: pim-rps
    __run_num__: 2
    __sls__: router.mcast.pim
    changes:
      diff: "@@ -44,6 +44,10 @@\n !\n ip multicast-routing\n !\n+router pim sparse-mode\n+
        \  ip pim rp-address 192.0.2.1 access-list a-mcast\n+   ip pim rp-address
        192.0.2.2 access-list b-mcast\n+!\n management api http-commands\n    no
        shutdown\n !"
    comment: ''
    duration: 2303.705
    name: pim-rps
    pchanges:
      diff: "@@ -44,6 +44,10 @@\n !\n ip multicast-routing\n !\n+router pim sparse-mode\n+
        \  ip pim rp-address 192.0.2.1 access-list a-mcast\n+   ip pim rp-address
        192.0.2.2 access-list b-mcast\n+!\n management api http-commands\n    no
        shutdown\n !"
    result: true
    start_time: '18:49:29.613658'
```

### Tests written?
No

### CODE REVIEW

**Comment by cachedout** [comment]
@bewing Could you clean up these two lint errors, please? https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/9752/violations/file/salt/utils/napalm.py/

**Review by mirceaulinic — CHANGES_REQUESTED** [review]
Two small requests:

Please add one `\n` after "To be changed:" when in test mode, e.g.:

```
device1:
----------
          ID: V4-FILTER-PDX-IN
    Function: netacl.managed
      Result: None
     Comment: To be changed: [edit firewall]
              +    family inet6 {
              +        /*
              +         ** $Id: V4-FILTER-PDX-IN $
              +         ** $Revision: 5 $
              +         **
              +         */
              +        filter my-filter {
              +            interface-specific;
              +            term my-other-term {
              +                from {
              +                    next-header tcp;
              +                    source-port 5678-5680;
              +                }
              +                then accept;
              +            }
              +        }
              +    }
              Configuration discarded.
     Started: 10:51:33.386009
    Duration: 9228.75 ms
     Changes:

Summary for device1
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
Total run time:   9.229 s
```

(just a cosmetic detail)

When in testing mode, the debug argument seems to be ignored -- it does not return the configuration loaded. (it returns correctly on `test=False` though).

File: salt/utils/napalm.py [hunk 1]
```diff
@@ -332,3 +332,68 @@ def func_wrapper(*args, **kwargs):
...
+            return ret
+        # Not test, changes were applied
+        ret.update({
+            'result': True,
+            'changes': {
+                'diff': loaded.get('diff', '')
+            },
+            'comment': loaded.get('comment', "Configuration changed!")
+        })
+        if debug:
```
Comment (mirceaulinic) [inline salt/utils/napalm.py]: On debug more, we should return the loaded config even when there was northing to change on the device. Otherwise you may wonder what have you generated that did not apply any changes.

File: salt/utils/napalm.py [hunk 2]
```diff
@@ -332,3 +332,68 @@ def func_wrapper(*args, **kwargs):
...
+            }
+        })
+        if test:
+            comment = "To be changed: {}\n{}".format(ret.get('pchanges', '').get('diff', ''),
+                                                     loaded.get('comment', ''))
+            ret.update({
+                'result': None,
+                'comment': comment
+            })
+            return ret
```
Comment (mirceaulinic) [inline salt/utils/napalm.py]: Probably that's why debug arg is ignored when in testing mode. Maybe we should check the debug arg somewhere at the begging.

**Comment by mirceaulinic** [comment]
Overall looks good, I have also tested this branch; there are some small details to be addressed and we're ready to go! Thanks @bewing!

File: salt/utils/napalm.py (hunk 1 above)
Comment (bewing) [inline salt/utils/napalm.py]: Do we want to revisit where we store the debug data?  I'm not sure "changes" is the appropriate key for this anymore.

Comment (mirceaulinic) [inline salt/utils/napalm.py]: Well, we only have three keys that are displayed: `result`, `comment` and `changes`. Out of them I think `changes` is the most appropriate.
We could put it in a separate key, but then it won't be displayed and using the raw outputter is very very hard to follow, so there will be little gain having it.
In other words, although I would be open to move it, we don't have other good options, in order to keel it useful.

Comment (bewing) [inline salt/utils/napalm.py]: So if testing, put it in pchanges, and if not testing, put it in pchanges and changes?

One of the goals was to NOT use `changes` if we are not actually changing anything on the device.  I would lean more towards stacking it into the comments field for now anyway, and perhaps long-term.

Comment (mirceaulinic) [inline salt/utils/napalm.py]: Yes, I agree with you.

How about the following: when in testing & debug mode together put also the rendered data in the comment. Would that make sense?

**Comment by mirceaulinic** [comment]
Looks good now - maybe one extra `\n` between "To be changed:" and "Debug loaded_config:"?
Somehow it looks too agglomerated:

```
device1:
----------
          ID: foo
    Function: netacl.managed
      Result: None
     Comment: To be changed:
              [edit firewall family inet]
              -     /*
              -      ** $Id: foo $
              -      ** $Date: 2017/03/16 $
              -      ** $Revision: 6 $
              -      **
              -      */
              +     /*
              +      ** $Id: foo $
              +      ** $Revision: 7 $
              +      **
              +      */
                    filter my-filter { ... }
              [edit firewall family inet filter my-filter term my-term from]
              -       source-port [ 123 389 161 ];
              +       source-address {
              +           1.2.3.4/32;
              +       }
              +       protocol [ udp tcp ];
              +       source-port [ 1235 1234 ];
              [edit firewall family inet filter my-filter term my-term then]
              -       count my-counter;
              -       syslog;
              -       accept;
              +       reject;
              [edit firewall family inet filter my-filter]
                     term my-term { ... }
              +      term my-other-term {
              +          from {
              +              protocol tcp;
              +              source-port 5678-5680;
              +          }
              +          then accept;
              +      }
              Debug loaded_config:
              firewall {
                  family inet {
                      replace:
                      /*
                      ** $Id: V4-FILTER-PDX-IN $
                      ** $Revision: 5 $
                      **
                      */
                      filter my-filter {
                          interface-specific;
                          term my-term {
                              from {
                                  source-address {
                                      1.2.3.4/32;
                                  }
                                  protocol [ udp tcp ];
                                  source-port [ 1235 1234 ];
                              }
                              then {
                                  reject;
                              }
                          }
                          term my-other-term {
                              from {
                                  protocol tcp;
                                  source-port 5678-5680;
                              }
                              then accept;
                          }
                      }
                  }
              }
              Configuration discarded.
     Started: 16:05:00.547824
    Duration: 10892.957 ms
     Changes:

Summary for device1
------------
Succeeded: 1 (unchanged=1)
Failed:    0
------------
Total states run:     1
Total run time:  10.893 s
```

**Comment by mirceaulinic** [comment]
Ah: when in debug more, we should have a key under the `pchanges` with the loaded config.

**Review by mirceaulinic — APPROVED** [review]
Thanks @bewing!

**Review by cachedout — APPROVED** [review]
Very clean. I like this a lot.
